### ISSUE TITLE
Build compiles but imports break when using optimization.splitChunks.chunks = 'all' with Module Federation

### ISSUE THREAD

**Issue by pratyushseq** [web-infra-dev/rspack#5332]
### System Info

  System:
    OS: macOS 13.6.3
    CPU: (8) arm64 Apple M1 Pro
    Memory: 100.59 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.6.0 - ~/Library/Caches/fnm_multishells/85041_1705309080052/bin/node
    Yarn: 1.22.19 - ~/Library/Caches/fnm_multishells/85041_1705309080052/bin/yarn
    npm: 8.13.2 - ~/Library/Caches/fnm_multishells/85041_1705309080052/bin/npm
  Browsers:
    Chrome: 120.0.6099.216
    Safari: 17.2.1

### Details

Add the following config in a module federation project, and the projects will compile, but the imports break in the actual app.

`optimization: {
  splitChunks: {
    chunks: 'all',
  }
}`

As discussed with @ahabhgk and @ScriptedAlchemy, opening this issue to keep track of future solutions

### Reproduce Steps

`yarn dev` with the config mentioned in any project with Module Federation

**Comment by Gillibald** [comment]
I had to explicitly set this to be able to load the remote:
```
chunkSplit: {
            strategy: 'split-by-module'
        }
```

otherwise the resulting `__webpack_exports__` was `undefined`

**Comment by Gillibald** [comment]
Coming back to this it even became worse for me. Now I am only able to load the remote by specifying `strategy: 'all-in-one'`

**Comment by pratyushseq** [comment]
@Gillibald i just left the whole `splitChunks` part out, let it fallback to the default `async` mode. That works reasonably well for me.

**Comment by ScriptedAlchemy** [comment]
This issue is caused by a conflict between webpack and chunk split with federation. To resolve this, you can use `enforce` on cacheGroups to override the splits. It is usually safer to use split-by-module or async, as initial module splits can cause the webpack exports to be wrapped in callbacks. This can result in the return value being undefined, rather than the container interface.

**Comment by Gillibald** [comment]
Thanks a lot. This fixes my issue. How can I achieve this with a pure `rsbuild` config without touching the `rspack` config override?

**Comment by xc2** [comment]
> Thanks a lot. This fixes my issue. How can I achieve this with a pure `rsbuild` config without touching the `rspack` config override?

@Gillibald would you please try out rsbuild 0.4.0

https://github.com/web-infra-dev/rsbuild/releases/tag/v0.4.0

cc @chenjiahan

**Comment by Gillibald** [comment]
I still need to specify:
```
rspack: {
            optimization: {
                splitChunks: {
                    chunks: 'async'
                }
            }
```
otherwise, the `__webpack_exports__` of the remote is `undefined`

**Comment by chenjiahan** [comment]
@Gillibald are you using the [moduleFederation.options](https://rsbuild.dev/config/module-federation/options) config?

**Comment by Gillibald** [comment]
rsbuild.config.ts
```
tools: {
        rspack: {
            optimization: {
                splitChunks: {
                    chunks: 'async'
                }
            },
            output: {
                publicPath: 'auto' //required otherwise chunk loading fails
            },
            plugins: [new container.ModuleFederationPlugin(federationConfig)]
        }
    }
})
```
rsbuild.config.mf.ts
```
const federationConfig: ModuleFederationPluginOptions = {
    name: 'hiwiki',
    exposes: {
        './a': './src/components/a',
        './b': './src/components/b',
        './c': './src/components/c',
    },
    remotes: {
        client: 'client@remoteEntry.js',
        lib:
            process.env.NODE_ENV === 'development'
                ? 'lib@http://localhost:3002/ogolib.js'
                : 'lib@ogolib.js'
    },
    filename: 'remoteEntry.js',
    shared: {
        react: {
            singleton: true,
            requiredVersion: dependencies['react']
        },
        'react-dom': {
            singleton: true,
            requiredVersion: dependencies['react-dom']
        },
        'react-singleton-context': {
            singleton: true,
            requiredVersion: dependencies['react-singleton-context']
        }       
    }
}

export default federationConfig
```

**Comment by Gillibald** [comment]
I am not using options here but I am manually adding the plugin so this might cause issues

**Comment by ScriptedAlchemy** [comment]
Yes using the rsbuild built-in capability allows to patch is accordingly so chunk split does not collide

**Comment by hardfist** [comment]
bump

**Comment by ScriptedAlchemy** [comment]
Should be fixed when i port hoisted runtime to rust

**Comment by hardfist** [comment]
close since it's fixed
